pkg/window/sorted_window_endtime.go

```diff
@@ -0,0 +1,276 @@
+/*
+Copyright 2022 The Numaproj Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package window
+
+import (
+	"sort"
+	"sync"
+	"time"
+)
+
+// SortedWindowListByEndTime is a thread safe list implementation, which is sorted by window end time
+// from lowest to highest.
+// All search operations are done using binary-search because it is an already sorted list.
+type SortedWindowListByEndTime struct {
+	windows []TimedWindow
+	lock    *sync.RWMutex
+}
+
+// NewSortedWindowListByEndTime implements a window list ordered by the end time. The Front/Head of the list will always have the smallest
+// element while the End/Tail will have the largest element (end time).
+func NewSortedWindowListByEndTime() *SortedWindowListByEndTime {
+	return &SortedWindowListByEndTime{
+		windows: make([]TimedWindow, 0),
+		lock:    &sync.RWMutex{},
+	}
+}
+
+// InsertBack inserts a window to the back of the list.
+func (s *SortedWindowListByEndTime) InsertBack(window TimedWindow) bool {
```

Comment:
`InsertFront` and `InsertBack` don't involve binary search operation, for some cases we need it. I prefer to keep these.
we can remove return types, since its not being used